Refactor `JsObject` to always be size 8

raskad opened the pull request:
This is currently a draft, since I think there are some important design decisions that need to be discussed.If you look at this, best is to review the changes commit by commit.

This PR contains two changes that can be best viewed separately.

1. [Refactor `JsObject` to always be a size 8 thin pointer (`Gc<T: Sized>`).](https://github.com/boa-dev/boa/commit/89c136ab2b0cbbbfdc7467a5b0f2c2c4ea72928b)
2. [Building on 1., remove the `Box` from the nan-boxed `JsObject` value.](https://github.com/boa-dev/boa/commit/8aa6d2af956a1ead9610c7e80c287d93ca9e0ba0)


## Motivation

I think the ideal representation for `JsObject` and `JsValue` should be a `no Drop` + `Copy` size 8 value. Currently we have a lot of drop and clone code in our runtime, caused by our ref-counting GC, as well as a lot of `Box` allocations from our current nan-box creation / clones. Hopefully we can get rid of the `Drop` and `Clone` requirements for `Gc` in the future. Then we could also move all other pointer types (`JsString`, `JsSymbol` and `JsBigInt`) to be `Gc` values, to enable `JsValue` to be `Copy`.

## Refactor `JsObject` to always be a size 8 thin pointer (`Gc<T: Sized>`)

Currently our `JsObject` type can be either typed (e.g. `JsObject<u64>`) or erased (e.g. `JsObject` == `JsObject<dyn NativeObject>)`. While there is only a `Gc` field on `JsObject`, the erased variant is a size 16 fat pointer, to store the Rust DST vtable for `dyn NativeObject`. The goal of this change is to make `JsObject` always a transparent size 8 thin pointer (`Gc`) to enable storing it directly as a nan-boxed value.

Here are the requirements for `JsObject` that I tried to keep:

* `JsObject` needs to be a size 8 thin pointer.
* Typed and erased variants should be one struct (erased being the default).
  * Possible alternative: `JsObject` could be a non generic struct and `JsObject::downcast` could return a typed struct. The positive of this approach would be, that access to the data T of the object can be controlled better, which would probably make everything safer. One drawback, that I currently see, is, that functions that should work on `JsObject<T>` would have to be either specific to one variant, or need to be implemented twice. I have not tried this approach, but I think in theory it should be possible.
* Any type that implements `NativeObject` should be able to be used as `T` for `JsObject` without any other user interactions (e.g. registering the type, manual unsafe vtable implementation, ...)

This lead me to the current design in this PR. Instead of erasing the object into a `dyn NativeObject`, the `TypeId` is stored in the object itself and used to check during casts. The default `T` is erased into a size 0 struct. I don't really like this approach, since one could access the erased struct. It might also be that some of this is inherently unsafe. Any ideas are welcome.


## Using the `JsObject` `Gc` pointer as the value for nan boxing.

At first this seems straight forward. But since there is no additional `Box` around the real value, it seems to me, that returning a reference to the `JsObject` from a nan boxed `JsValue` is impossible.
This alters the `JsValue::as_object` API to return a `Option<JsObject>` instead of `Option<&JsObject>`, which cascades out into some breaking changes in our `JsValue` APIs and some arguably less ergonomic usage patterns.
I think we should break the APIs, since when we move all pointer types and `JsValue` to being `Copy`, we basically don't want to work with references to those types anyway and always take them as values. This would of course mean a lot of breaking changes in our APIs, but I think it would be the logical thing to do.

## Benchmarks

The current changes obviously don't change anything related to the GC, but removing the additional `Box` for nan-boxed `JsObject`s seems to already make a difference. Local benchmark results:

Before

```
RESULT Richards 221
RESULT DeltaBlue 218
RESULT Crypto 208
RESULT RayTrace 485
RESULT EarleyBoyer 591
RESULT RegExp 104
RESULT Splay 829
RESULT NavierStokes 485
SCORE 324
```

After

```
RESULT Richards 276
RESULT DeltaBlue 281
RESULT Crypto 226
RESULT RayTrace 587
RESULT EarleyBoyer 721
RESULT RegExp 110
RESULT Splay 1039
RESULT NavierStokes 513
SCORE 380
```

hansl commented:
> This alters the JsValue::as_object API to return a Option<JsObject> instead of Option<&JsObject>, which cascades out into some breaking changes in our JsValue APIs and some arguably less ergonomic usage patterns.

I think this is a very good change, and will inevitably be done for the other "pointer" types (`BigInt`, `String`, etc). If we can make `JsObject` Copy, there shouldn't be any references to it anywhere anymore. Should improve benchmarks even further. I like this a lot.

jedel1043 commented on core/engine/src/object/jsobject.rs:
```diff
@@ -42,16 +42,22 @@ pub type Ref<'a, T> = boa_gc::GcRef<'a, T>;
 #[boa_gc(unsafe_no_drop)]
-pub struct JsObject<T: NativeObject + ?Sized = dyn NativeObject> {
-    inner: Gc<VTableObject<T>>,
+pub struct JsObject<T: NativeObject + ?Sized = ErasedObjectData> {
```
I'm pretty sure we do something fairly similar to this in the garbage collector code

HalidOdat replied:
Hmmm... this may not be needed, when we construct a `Gc<T>` we keep the trace and finalize function of `T` in the gc's vtable. And calling [`trace` on `Gc<T>` erases its type when it's queued](https://github.com/boa-dev/boa/blob/main/core/gc/src/pointers/gc.rs#L166-L171), and when we trace we [call the vtable's trace](https://github.com/boa-dev/boa/blob/main/core/gc/src/trace.rs#L49-L53).

So doing this should be safe :)

jedel1043 replied:
Ohhh true, that's really nice to have actually! We should definitely add a comment explaining this then, since it's pretty hard to miss

HalidOdat replied:
Yes, if we some additional type ID data in the `Gc`'s `VTable`, we can enable easy `downcast` functionality on the `Gc` pointer level, that way we may be able to eliminate the type ID in the object vtable. :)

jedel1043 commented:
> That was also something I though about and I think it could be removed now, because it was only there to allow for the erased dyn NativeObject, correct?

Yes. We could technically enable something like `Object<[u8]>` if we left the `?Sized` requirement, but I think `NativeObject` requires `Sized`, so it wouldn't be useful at all to have that if we're removing `dyn NativeObject`.

> It's because of the alignment of Object. If we don't box T, the alignment would depend on the alignment of T. That would mean, that we cannot safely access fields on Object while it is erased (at least as far as I understand).

Maybe the field is already aligned to 8 bytes thanks to the rest of the fields? We could also force `#[repr(C, align(64)]` and add padding so that the data field is always 8 byte aligned.

raskad commented:
> Yes. We could technically enable something like `Object<[u8]>` if we left the `?Sized` requirement, but I think `NativeObject` requires `Sized`, so it wouldn't be useful at all to have that if we're removing `dyn NativeObject`.

Sounds good!
 
> Maybe the field is already aligned to 8 bytes thanks to the rest of the fields? We could also force `#[repr(C, align(64)]` and add padding so that the data field is always 8 byte aligned.

Can we force the struct into a specific alignment? The reference for repr C documents, that the alignment is always the alignment of the most aligned field. When I test the following code, `Object` is still 2^29 aligned:

```rust
#[repr(align(536870912))]
struct Inner {
    field: u8,
}

#[repr(C, align(64))]
struct Object<T> {
    field: T,
}

```

jedel1043 commented:
> Can we force the struct into a specific alignment? The reference for repr C documents, that the alignment is always the alignment of the most aligned field. When I test the following code, Object is still 2^29 aligned:

Oh true, I think I got confused by how `#[repr(C)]` works. Thinking it through a bit more, I think to avoid the `Box` we just need to add ``#[repr(C)]`, because in that case the automatic padding to align `data` would be just before `data` itself, right? Thus, the previous fields won't change their offset, making it them safe to access.

jedel1043 commented:
> Possible alternative: JsObject could be a non generic struct and JsObject::downcast could return a typed struct. The positive of this approach would be, that access to the data T of the object can be controlled better, which would probably make everything safer. One drawback, that I currently see, is, that functions that should work on JsObject<T> would have to be either specific to one variant, or need to be implemented twice. I have not tried this approach, but I think in theory it should be possible.

Note that most of our methods are only implemented for `JsObject<dyn NativeObject>`, because most of the JS functions only take the erased version of `JsObject`. There are only a handful of methods implemented for `JsObject<T>`, so going with this approach shouldn't be that much of a problem I think.

jedel1043 commented:
@raskad Ahh yeah, I didn't consider the alignment of `VTableObject`, and any other struct that wraps `Object.

raskad commented:
Got an update on this. I rebased this on the changes from #4291 and used `GcErased` in `JsObject`. @HalidOdat I needed to switch the `*cast` functions on `GcErased` to use take a value instead of a reference, to avoid clones of `Gc<T>`. For most other scenarios, I added a `downcast_ref` function. What do you think abou this?

Building on you comment @jedel1043

> Note that most of our methods are only implemented for `JsObject<dyn NativeObject>`, because most of the JS functions only take the erased version of `JsObject`. There are only a handful of methods implemented for `JsObject<T>`, so going with this approach shouldn't be that much of a problem I think.

I added a `JsObjectTyped<T>` struct that can be used for the downcasted `JsObjects`. You where totally correct, there where only a couple of methods, that I just copied for now. The biggest breaking change would be, that users would need to change their usages. What do you all think (you can see our internal usage changes to guage the effort).

HalidOdat commented on core/engine/src/object/jsobject.rs:
```diff
@@ -1039,13 +1104,154 @@ impl<T: NativeObject + ?Sized> Debug for JsObject<T> {
+        reason = "can only use `ptr::fn_addr_eq` on rustc 1.85"
+    )]
+    pub fn is_callable(&self) -> bool {
+        self.inner.vtable.__call__ != ORDINARY_INTERNAL_METHODS.__call__
```
nitpick (non-blocking): It doesn't change much but we could use the [`fn_addr_eq()`](https://doc.rust-lang.org/std/ptr/fn.fn_addr_eq.html). (same for `is_constructor()` below)

HalidOdat commented on core/engine/src/object/jsobject.rs:
```diff
@@ -865,29 +867,92 @@ impl<T: NativeObject + ?Sized> JsObject<T> {
+    /// Note that the returned object will not be erased to be convertible to a
+    /// `JsValue`. To erase the pointer, call [`JsObjectTyped::upcast`].
+    pub fn new<O: Into<Option<JsObject>>>(root_shape: &RootShape, prototype: O, data: T) -> Self
+    where
+        T: Sized,
```
`T` is always `Sized` because of the `impl` constraint so we can remove this (applies to all functions in this `impl` block) :)

```suggestion
```

HalidOdat commented on core/engine/src/object/jsobject.rs:
```diff
@@ -24,38 +24,64 @@ use crate::{
+}
+
+/// Garbage collected typed `Object`.
+#[derive(Trace, Finalize)]
+#[boa_gc(unsafe_no_drop)]
+pub struct JsObjectTyped<T: NativeObject> {
+    pub(crate) inner: Gc<VTableObject<T>>,
 }
```
```suggestion
#[derive(Trace, Finalize)]
#[boa_gc(unsafe_no_drop)]
pub struct JsObjectTyped<T: NativeObject = ErasedObjectData> {
    pub(crate) inner: Gc<VTableObject<T>>,
}
```

Couldn't we have a default parameter here and it would eliminate the need for two `JsObject` types.

raskad replied:
Yes, we could keep it like this and avoid `JsObjectTyped`. I just wanted to try it out like this to leverage `GcErased`. As per our discussion, I personally don't think there is a big benefit to splitting the erased and typed variants, so unless someone can think of a big benefit, I would also be for keeping the default paramter.

raskad replied:
> We could still use GcErased in JsObject (if we wanted). It would just requires having a small method that we used to access inner as VTableObject<T> (by casting with GcErased::cast).

And we would need a `PhantomData<T>`, right? That would be my reason just to use `Gc<T>` directly, since the code would be easier.

HalidOdat replied:
> And we would need a `PhantomData<T>`, right? That would be my reason just to use `Gc<T>` directly, since the code would be easier.

Yes, phantom data is required to appease the compiler from the warning of unused `T`. Yes, using `Gc<T>` will be simpler :smile:

HalidOdat reviewed (commented):
Really nice work! Really like the performance boost! Check my comments if they make any sense :smile:

HalidOdat commented on core/engine/src/object/jsobject.rs:
```diff
@@ -92,9 +119,11 @@ impl JsObject {
-        Self {
-            inner: coerce_gc(gc),
-        }
+        // SAFETY: The pointer is guaranteed to be valid because we just created it.
+        // `VTableObject<ErasedObjectData>` and `VTableObject<T>` have the same size and alignment.
+        let inner = unsafe { Gc::cast_unchecked::<ErasedVTableObject>(gc) };
+
+        JsObject { inner }
```
```suggestion
        JsObject { inner: gc }.upcast()
```

We could eliminate the unsafe by using the `upcast` method which already has the unsafe code.

HalidOdat commented on core/engine/src/object/jsobject.rs:
```diff
@@ -138,17 +167,19 @@ impl JsObject {
-        Self {
-            inner: coerce_gc(gc),
-        }
+        // SAFETY: The pointer is guaranteed to be valid because we just created it.
+        // `VTableObject<ErasedObjectData>` and `VTableObject<T>` have the same size and alignment.
+        let inner = unsafe { Gc::cast_unchecked::<ErasedVTableObject>(gc) };
+
+        JsObject { inner }
```
Same here :)

HalidOdat commented on core/engine/src/object/jsobject.rs:
```diff
@@ -177,9 +208,11 @@ impl JsObject {
-        Self {
-            inner: coerce_gc(gc),
-        }
+        // SAFETY: The pointer is guaranteed to be valid because we just created it.
+        // `VTableObject<ErasedObjectData>` and `VTableObject<T>` have the same size and alignment.
+        let inner = unsafe { Gc::cast_unchecked::<ErasedVTableObject>(gc) };
+
+        JsObject { inner }
```
Same here :)

HalidOdat commented on core/engine/src/object/jsobject.rs:
```diff
@@ -188,25 +221,9 @@ impl JsObject {
-            let inner = unsafe { Gc::from_raw(ptr) };
+        if self.is::<T>() {
+            // SAFETY: We have verified that the object is of type `T`, so we can safely cast it.
+            let inner = unsafe { Gc::cast_unchecked::<VTableObject<T>>(self.inner) };
```
Consider using the `JsObject` downcast method :)

```suggestion
            let inner = unsafe { self.downcast_unchecked::<T>() };
```

HalidOdat commented on core/engine/src/object/jsobject.rs:
```diff
@@ -620,88 +651,24 @@ Cannot both specify accessors and a value or writable attribute",
-
-        if obj.is::<ArrayBuffer>() {
-            drop(obj);
+        if self.is::<ArrayBuffer>() {
             // SAFETY: We have verified that the inner data of `self` is of type `ArrayBuffer`.
             return Ok(BufferObject::Buffer(unsafe {
                 self.downcast_unchecked::<ArrayBuffer>()
             }));
         }
```
I think we can remove the unsafe here by chaining downcast calls :)

```suggestion
        match self.downcast::<ArrayBuffer>() {
            Ok(buffer) => return Ok(BufferObject::Buffer(buffer)),
            Err(object) => object.downcast::<SharedArrayBuffer>().map(BufferObject::SharedBuffer)
        }
```
